### PR TITLE
feat: implement dotenv-based secret management system

### PR DESCRIPTION
**Pull request by aaronsteers**
# feat: implement dotenv-based secret management system

## Summary

This PR implements a comprehensive dotenv-based secret management system for connector-builder-mcp that allows users to manage secrets in .env files without exposing actual secret values to the LLM. The system provides MCP tools for managing secrets and automatically hydrates connector configurations with secrets during execution.

**Key Features:**
- **File-based secret management**: Uses .env files with configurable paths via environment variable or MCP tool
- **Naming convention mapping**: Environment variables map to config paths (e.g., `CREDENTIALS_PASSWORD` → `credentials.password`)
- **MCP tools**: `set_dotenv_path`, `list_secrets`, `add_secret_stub`, `get_secrets_file_path_for_user`
- **Automatic config hydration**: Seamlessly injects secrets into connector configs before execution
- **Security-first**: LLM never sees actual secret values, only metadata about what secrets exist

**Implementation Details:**
- Added `python-dotenv>=1.0.0` dependency
- Created `_secrets.py` module with global state management for dotenv file path
- Integrated config hydration into all three connector builder operations: `validate_manifest`, `execute_stream_test_read`, and `get_resolved_manifest`
- Comprehensive test suite with 18 test cases covering all functionality

## Review & Testing Checklist for Human

- [ ] **End-to-end workflow test**: Use MCP client to set dotenv path, add secret stubs, populate with real values, and verify connector operations work correctly
- [ ] **Naming convention verification**: Test various environment variable patterns (`API_KEY`, `CREDENTIALS_PASSWORD`, `OAUTH_CLIENT_SECRET`) map correctly to config paths
- [ ] **Config hydration integration**: Verify existing connector operations still work and secrets are properly injected without breaking functionality
- [ ] **Tool registration**: Confirm all four new MCP tools are accessible and work as expected
- [ ] **File handling edge cases**: Test with various file paths, missing directories, permissions issues, and concurrent access scenarios

**Recommended Test Plan:**
1. Start MCP server with `poe server` and connect with MCP client
2. Use `set_dotenv_path` to configure a test .env file
3. Use `add_secret_stub` to add placeholder secrets
4. Manually edit the .env file to add real secret values
5. Run connector validation/testing operations and verify secrets are hydrated correctly
6. Test `list_secrets` shows secret keys without exposing values

---

### Diagram

```mermaid
%%{ init : { "theme" : "default" }}%%
graph TD
    Client["MCP Client"]
    Server["connector_builder_mcp/server.py"]:::context
    ConnectorBuilder["_connector_builder.py"]:::major-edit
    Secrets["_secrets.py"]:::major-edit
    EnvFile[".env file"]
    Tests["tests/test_secrets.py"]:::major-edit
    PyProject["pyproject.toml"]:::minor-edit
    Util["_util.py"]:::minor-edit
    
    Client --> Server
    Server --> ConnectorBuilder
    ConnectorBuilder --> Secrets
    Secrets --> EnvFile
    Tests --> Secrets
    
    ConnectorBuilder -.->|"hydrate_config()"| Secrets
    Secrets -.->|"load_secrets()"| EnvFile
    Client -.->|"set_dotenv_path<br/>list_secrets<br/>add_secret_stub"| Secrets

    subgraph Legend
        L1[Major Edit]:::major-edit
        L2[Minor Edit]:::minor-edit
        L3[Context/No Edit]:::context
    end

    classDef major-edit fill:#90EE90
    classDef minor-edit fill:#87CEEB
    classDef context fill:#FFFFFF
```

### Notes

- The `_env_var_to_path` function contains complex logic for mapping environment variable names to config paths - this is a critical piece that should be thoroughly tested
- Global state management with `_current_dotenv_path` could potentially have concurrency issues in multi-threaded scenarios
- Config hydration happens automatically before every connector operation, so any bugs here could break existing functionality
- The implementation assumes users will follow the naming convention correctly - consider adding validation or better error messages for misnamed variables

**Link to Devin run**: https://app.devin.ai/sessions/eab82b57cce84fb8aeb33b11e2aabea0  
**Requested by**: AJ Steers (@aaronsteers)

### CODE REVIEW

File: connector_builder_mcp/_secrets.py [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+
+        secret_names = [name for name, _ in secrets_to_add]
+        return f"Added {added_count} secret stub(s) to {secrets_file}: {', '.join(secret_names)}. Please set the actual values."
+
+    except Exception as e:
+        logger.error(f"Error adding secret stubs: {e}")
+        return f"Error adding secret stubs: {str(e)}"
+
+
+def _extract_secrets_from_manifest(manifest: dict[str, Any]) -> list[tuple[str, str]]:
```
Comment (aaronsteers) [inline connector_builder_mcp/_secrets.py:273]: The function name raises alarm unnecessarily. Should be "exact secrets names from manifest". Longer but necessary.

File: connector_builder_mcp/_secrets.py [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+        current = obj
+        for key in path[:-1]:
+            if key not in current:
+                current[key] = {}
+            elif not isinstance(current[key], dict):
+                return
+            current = current[key]
+        current[path[-1]] = value
+
+    def _env_var_to_path(env_var: str) -> list[str]:
```
Comment (aaronsteers) [inline connector_builder_mcp/_secrets.py:133]: No, this isn't possible. Don't try. Only resolve the other way around. Start with the full domain of settings (secrets or not) that you might be provided, and then you can pair them and find out where they go.

File: connector_builder_mcp/_secrets.py [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+    config_paths: Annotated[
+        list[str] | None,
+        Field(
+            description="List of config paths like ['credentials.password', 'oauth.client_secret']"
+        ),
+    ] = None,
+    secret_key: Annotated[
+        str | None, Field(description="Single secret key to add (legacy mode)")
+    ] = None,
+    description: Annotated[str, Field(description="Description for the secret(s)")] = "",
```
Comment (aaronsteers) [inline connector_builder_mcp/_secrets.py:199]: Remove the legacy and remove all secret descriptions. Secrets don't need descriptions.

File: connector_builder_mcp/_connector_builder.py [hunk 4]
```diff
@@ -161,6 +163,7 @@ def execute_stream_test_read(
     logger.info(f"Testing stream read for stream: {stream_name}")
 
     try:
+        config = hydrate_config(config, dotenv_path=None)
```
Comment (aaronsteers) [inline connector_builder_mcp/_connector_builder.py:166]: The tool needs to accept the (optional) dotenv_path from the LLM.

File: connector_builder_mcp/_secrets.py [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+class SecretInfo(BaseModel):
+    """Information about a secret without exposing its value."""
+
+    key: str
+    is_set: bool
+    description: str = ""
```
Comment (aaronsteers) [inline connector_builder_mcp/_secrets.py:24]: Get rid of the secret description field everywhere.

File: tests/test_secrets.py
Comment (aaronsteers) [inline tests/test_secrets.py:1]: Audit your test implementation and report back in a comment:
- Remove any useless tests.
- Alarm on any inadvertently suppressed signals.
- Alarm on any wastful redundancies.

Post a comment before addressing. If all is fine, great. If some can be improved, first comment, then fix, then comment again.

File: connector_builder_mcp/_secrets.py [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+) -> str:
+    """Get the absolute path to the specified dotenv file for user reference.
+
+    Args:
+        dotenv_path: Path to the .env file
+
+    Returns:
+        Absolute path to the dotenv file
+    """
+    return str(Path(dotenv_path).absolute())
```
Comment (aaronsteers) [inline connector_builder_mcp/_secrets.py:253]: You were supposed to delete this. I'm deleting. Pull the latest when I'm done.

```suggestion
```

File: connector_builder_mcp/_connector_builder.py [hunk 7]
```diff
@@ -146,6 +147,10 @@ def execute_stream_test_read(
         int,
         Field(description="Maximum number of records to read", ge=1, le=1000),
     ] = 10,
+    dotenv_path: Annotated[
+        str | None,
```
Comment (aaronsteers) [inline connector_builder_mcp/_connector_builder.py:151]:
```suggestion
        Path | None,
```
